components/schematics/ng-add/ast/update-ngmodule-decorators.ts

```diff
@@ -0,0 +1,209 @@
+  let decorators = findNodes(
+    sourceFile,
+    ts.SyntaxKind.Decorator,
+  ) as ts.Decorator[];
+
+  decorators.forEach((decorator: ts.Decorator) => {
+    // check if the found decorator is the NgModule decorator and
+    // if its class is exported.
+    if (
+      ts.isCallExpression(decorator.expression) &&
+      ts.isIdentifier(decorator.expression.expression) &&
+      decorator.expression.expression.text === 'NgModule' &&
+      hasExportModifier(decorator.parent)
+    ) {
+      let updatedProperties: ts.ObjectLiteralExpression | undefined;
+      // loop through the object properties of the decorator to get the imports
+      decorator.expression.arguments.forEach(
+        (argument: ts.ObjectLiteralExpression) => {
+          const prop = argument.properties.find(
+            (property: ts.PropertyAssignment) => {
+              const name = getSymbolName(property);
+              if (name === propertyType) {
+                return true;
+              }
+              return false;
+            },
+          ) as ts.PropertyAssignment;
+
+          if (prop && propertyValue) {
+            // if we found the property append the value to the array if it is an array
+            if (ts.isArrayLiteralExpression(prop.initializer)) {
+              prop.initializer = ts.createArrayLiteral(
+                [...prop.initializer.elements, propertyValue],
+                true,
+              );
+            }
+          } else {
+            const propertiesArr: any[] = [];
+            if (propertyValue) {
+              propertiesArr.push(...argument.properties);
+              propertiesArr.push(
+                createPropertyAssignment(propertyType, propertyValue),
+              );
+            } else {
+              // if the propertyValue is undefined then we want to delete it!
+              argument.properties.forEach((property: ts.PropertyAssignment) => {
+                const name = getSymbolName(property);
+                // only push the properties where the name does not match the provided type
+                if (name !== propertyType) {
+                  propertiesArr.push(property);
+                }
+              });
+            }
+            // if the property does not exist in the NgModule we have to create it and append it!
+            updatedProperties = ts.createObjectLiteral(propertiesArr, true);
+          }
+        },
+      );
+
+      // if we have updated properties we have to recreate the call expression
+      if (updatedProperties) {
+        decorator.expression = ts.createCall(
+          ts.createIdentifier('NgModule'),
+          undefined,
+          [updatedProperties],
+        );
+      }
+    }
+  });
+  // const util = require('util')
+
+  //console.log(sourceFile);
+  // console.log(util.inspect(decorators[0]));
+
+  return sourceFile;
+}
+
+function findNodes(
+  node: ts.Node,
+  kind: ts.SyntaxKind,
+  max: number = Infinity,
+): ts.Node[] {
+  const arr: ts.Node[] = [];
```

Comment:
Could you add some comments to the code of this function. It's rather hard to grasp how it loops through to find the nodes.